### PR TITLE
Introduce Activity.IsStopped Property

### LINKED ISSUE #63353: [API Proposal]: Activity.IsStopped  (functionality required to conform to OpenTelemetry specification)
### Background and motivation

As far as I know, the Activity class is meant to provide all the functionality required by the OpenTelemetry tracing specification for the concept that is called "Span" in OTel terms.

However, I could not find a way to get the following from Activity:

>  It must also be able to reliably determine whether the Span has ended (some languages might implement this by having an end timestamp of null, others might have an explicit hasEnded boolean).

-- https://github.com/open-telemetry/opentelemetry-specification/blob/v1.8.0/specification/trace/sdk.md#additional-span-interfaces

At first, I thought one could check if `Activity.Duration != TimeSpan.Zero`, but while this is indeed true for all stopped/ended activities, even activities that are still running can have a non-zero duration if `SetEndTime` was called but `Stop`/`Dispose` was not.

In some situations, one could could use the callback from the Source and maintain a ConditionalWeakTable to record whether an Activity has stopped, but this is rather cumbersome, slow and not possible if you do not have control over the activities' source.

It seems that Activity already has this implemented as private functinality as a property `IsFinished`: https://github.com/dotnet/runtime/blob/v6.0.1/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L1301-L1303

Exposing  the existing name IsFinished would also be an acceptable IMHO (I think the most OTel-like name would be HasEnded, but since End is called Stop, my primary suggestion is IsStopped).

The interface in OpenTelemetry Java is here https://github.com/open-telemetry/opentelemetry-java/blob/v1.9.1/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/ReadableSpan.java#L64-L71 and here https://github.com/open-telemetry/opentelemetry-java/blob/v1.9.1/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/data/SpanData.java#L117-L122

CC @open-telemetry/dotnet-approvers

### API Proposal

```C#
namespace System.Diagnostics.Generic
{
    public class Activity: IDisposable
    {
        // ...
        public bool IsStopped { get; }
        // ...
    }
}
```

### API Usage

```C#
var a = new Activity("foo");
Debug.Assert(!a.IsStopped);
a.Start();
Debug.Assert(!a.IsStopped);
a.Stop();
Debug.Assert(a.IsStopped);
```

In an ActivityListener.ActivityStopped callback, `IsStopped` of the notifying Activity must also be true to conform to OTel semantics.

### Alternative Designs

If `SetEndTime` would always unconditionally `Stop` the activity, the check for nonzero `Duration` could be used instead of using a separate property. However, that would probably be an unacceptable breaking change.

### Risks

None known. This proposal would add a new property exposing existing functionality that is required as an API surface by the OpenTelemetry specification.

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/63353

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @Oberon00 @reyang @cijothomas

**Review by reyang — APPROVED** [review]

**Review by noahfalk — APPROVED** [review]

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 1]
```diff
@@ -784,6 +784,11 @@ public ActivityTraceId TraceId
         /// </summary>
         public bool IsAllDataRequested { get; set;}
 
+        /// <summary>
+        /// Indicate if the this Activity object is stopped
+        /// </summary>
+        public bool IsStopped { get => IsFinished; }
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:790]: Why not just rename IsFinished to IsStopped?

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:790]: No reason other than less code change :-) I'll apply this suggestion.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 2]
```diff
@@ -1298,18 +1298,21 @@ private bool W3CIdFlagsSet
             get => (_w3CIdFlags & ActivityTraceFlagsIsSet) != 0;
         }
 
-        private bool IsFinished
+        /// <summary>
+        /// Indicate if the this Activity object is stopped
```
Comment (danmoseley) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1302]:
```suggestion
        /// Indicates whether this Activity object is stopped
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 3]
```diff
@@ -1298,18 +1298,21 @@ private bool W3CIdFlagsSet
             get => (_w3CIdFlags & ActivityTraceFlagsIsSet) != 0;
         }
 
-        private bool IsFinished
+        /// <summary>
+        /// Indicates whether this Activity object is stopped
```
Comment (Oberon00) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1302]: Maybe document that this is already true in the ActivityStopped listener (assuming and hoping this is the case). 
Could also be added to the unit test.

**Review by Oberon00 — APPROVED** [review]

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs (hunk 3 above)
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:1302]:
```suggestion
        /// Indicates whether this <see cref="Activity"/> object is stopped.
```

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs [hunk 4]
```diff
@@ -320,6 +319,31 @@ public void StartStop()
             Assert.Null(Activity.Current);
         }
 
+        /// <summary>
+        /// Tests Activity.IsStopped
+        /// </summary>
+        [Fact]
+        public void IsStoppedTest()
+        {
+            var activity = new Activity("activity");
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs:328]:
```suggestion
            using var activity = new Activity("activity");
```
?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs (hunk 4 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs:328]: Dispose wouldn't add anything here as I am manually stopping the Activity. What is the value of having `using`?

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivityTests.cs:328]: If an assert fails before you call stop causing an exception to be thrown and the activity not disposed.  You know the impact best, though, so up to you.

**Review by cijothomas — APPROVED** [review]

**Pull request merged**